File: reserve/Lab_rc4/exp.py
# 定义一个RC4解密函数，它接受密钥和密文作为参数，然后返回明文
def rc4_decrypt(key, ciphertext):
    # 初始化S盒，它是一个256字节的列表，用来存储0到255的互不重复的元素
    S = list(range(256))
    # 初始化j，它是一个整数，用来记录S盒的交换位置
    j = 0
    # 根据密钥打乱S盒，这个过程叫做KSA（the Key-Scheduling Algorithm）
    for i in range(256):
        # j等于j加上S盒的第i个元素和密钥的第i个字符的ASCII码，再模256
        j = (j + S[i] + ord(key[i % len(key)])) % 256
        # 交换S盒的第i个元素和第j个元素
        S[i], S[j] = S[j], S[i]
    # 初始化i和j，它们都是整数，用来记录S盒的交换位置
    i = 0
    j = 0
    # 初始化out，它是一个空列表，用来存储解密后的字符
    out = []
    # 遍历密文的每个字符，这个过程叫做PRGA（the Pseudo-Random Generation Algorithm）
    for char in ciphertext:
        # i等于i加1再模256
        i = (i + 1) % 256
        # j等于j加上S盒的第i个元素再模256
        j = (j + S[i]) % 256
        # 交换S盒的第i个元素和第j个元素
        S[i], S[j] = S[j], S[i]
        # 生成一个密钥流的字节，它等于S盒的第i个元素和第j个元素之和再模256对应的S盒的元素
        k = S[(S[i] + S[j]) % 256]
        # 将密文的字符和密钥流的字节进行异或运算，得到解密后的字符，并添加到out列表中
        out.append(chr(ord(char) ^ k))
    # 将out列表中的所有字符拼接起来，形成一个字符串，并返回
    return ''.join(out)

# 调用RC4解密函数，传入密钥和密文，得到明文
key = b'Scr1w666' # 这是一个示例密钥，您可以根据您的情况修改
eventext = b'\x07\x9B\xE8\xAC\x68\xAB\xAF\x3E\xC4\x66\xD1\x0E\x88\x06\x8A\x11' # 这是一个示例密文，您可以根据您的情况修改
oddtext  = b'\x93\xB1\x33\xA3\xC5\x40\x3B\xB5\xF9\x64\xDF\xEB\xD7\xA8\x24\x88'

eventext = rc4_decrypt(key, eventext)
oddtext = rc4_decrypt(key, oddtext)
# 打印明文
print(eventext)
print(oddtext)
#\x07\x93\x9B\xB1\xE8\x33\xAC\xA3\x68\xC5\xAB\x40\xAF\x3B\x3E\xB5\xC4\xF9\x66\x64\xD1\xDF\x0E\xEB\x88\xD7\x06\xA8\x8A\x24\x11\x88

File: reserve/Lab_rc4/exp.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
unsigned int* sub_401410(int s, int a2, int key) {
    unsigned int* result = (unsigned int*)malloc(256 * sizeof(unsigned int));
    int v4 = 0;

    for (int i = 0; i <= 255; ++i) {
        result[i] = i;
        result[i + a2 / 4] = result[i % 8 + key / 4];
    }

    for (int j = 0; j <= 255; ++j) {
        v4 = (unsigned char)(v4 + result[j + s / 4] + result[j + a2 / 4]);
        result[j + s / 4] ^= result[v4 + s / 4];
        result[v4 + s / 4] ^= result[j + s / 4];
        result[j + s / 4] ^= result[v4 + s / 4];
    }

    return result;
}
int sub_401563(int *a1, int *a2, int a3)
{
  int result; // eax
  int i; // [esp+4h] [ebp-10h]
  int v5; // [esp+8h] [ebp-Ch]
  int v6; // [esp+Ch] [ebp-8h]

  v6 = 0;
unsigned char lowByte = LOBYTE(v5);
lowByte = 0;
  for ( i = 0; ; ++i )
  {
    result = i;
    if ( i >= a3 )
      break;
    v6 = (v6 + 1) % 256;
    v5 = (*(4 * v6 + a1) + v5);
    *(4 * v6 + a1) ^= *(4 * v5 + a1);
    *(4 * v5 + a1) ^= *(4 * v6 + a1);
    *(4 * v6 + a1) ^= *(4 * v5 + a1);
    *(4 * i + a2) ^= *(4 * (*(4 * v6 + a1) + *(4 * v5 + a1))
                                         + a1) ^ (i + 60);
  }
  return result;
}
int main(int argc, char const *argv[])
{
    int input[32]={0x7,0x93,0x9b,0xb1,0xe8,0x33,0xac,0xa3,0x68,0xc5,0xab,0x40,0xaf,0x3b,0x3e,0xb5,0xc4,0xf9,0x66,0x64,0xd1,0xdf,0xe,0xeb,0x88,0xd7,0x6,0xa8,0x8a,0x24,0x11,0x88};
    char v4[8] = "Scr1w666";
    int *v5 = malloc(120);
    int *v6 = malloc(120);
    int *v9 = malloc(1024);
    void *v8 = malloc(1024);
    memset(v9,0,256);
    memset(v8,0,256);

    memset(v6, 0, 30);
    memset(v5, 0, 30);

    for (int i = 0; i <= 15; ++i)
    {
        *( i + v6) = *(2 * i + input);//偶数项
        *( i + v5) = *(2 * i + 1 + input);//奇数项
    }
sub_401410(v9, v8, v4);
sub_401563(v9, v6, 16);
sub_401563(v9, v5, 16);
printf("%s",v6);
printf("%s",v5);

return 0;
}

